🦋 Site Usage Rules with PF4

josemigallas opened the pull request:
[THREESCALE-9949: Site Usare Rules](https://issues.redhat.com/browse/THREESCALE-9949)

Also fixes [THREESCALE-9833: Misaligned text for changing account and service plans](https://issues.redhat.com/browse/THREESCALE-9833)

### Before

![Screenshot 2023-07-14 at 12 40 44](https://github.com/3scale/porta/assets/11672286/1871aa64-dfab-48ae-bf9b-c021e0d324b3)-
![Screenshot 2023-07-14 at 12 40 49](https://github.com/3scale/porta/assets/11672286/a286966c-396a-47ce-b31f-5f2ee2051611)


### After
<img width="1298" alt="Screenshot 2023-08-01 at 17 13 19" src="https://github.com/3scale/porta/assets/11672286/ecafb7ae-9f62-4f00-a588-a1f1f025c8bc">
<img width="1298" alt="Screenshot 2023-08-01 at 17 11 58" src="https://github.com/3scale/porta/assets/11672286/3bb4bff4-b28f-407a-973c-4f6b08b2c09a">

mayorova commented:
I've seen a couple of issues so far:
![usage_rules](https://github.com/3scale/porta/assets/1270328/1aa1d8db-7bc9-4a6e-8c98-32765204fa75)

1. weird hint on "Developers are allowed sign up themselves"
(LOL, also "to" is missing here)
Can be fixed by adding `formtastic.hints.settings.signups_enabled: ""` to i18n

2. link is broken for "Account Approval Required" hint
(you need to have some SSO configured to make this part visible)
Also... this label is not taken from i18n (hence, this camel case, I guess), it shows as 
```<span class="translation_missing" title="translation missing: en.formtastic.labels.settings.account_approval_required">Account Approval Required</span>``` in the HTML

mayorova commented on app/views/sites/usage_rules/edit.html.slim:
```diff
@@ -1,60 +1,81 @@
+            div class="pf-c-form__group-control"
+              = render 'settings_check', setting: 'account_approval_required',
+                                         checked: settings.account_approval_required,
+                                         disabled: true,
```
missing handing of the `disabled` property in the `settings_check` partial?...

josemigallas replied:
Oops, I forgot

mayorova commented on app/views/sites/usage_rules/edit.html.slim:
```diff
@@ -1,60 +1,81 @@
+  div class="pf-c-card__body"
+    = semantic_form_for @settings, url: admin_site_usage_rules_path, html: { class: "pf-c-form pf-m-limit-width" } do |form|
+      section class="pf-c-form__section" role="group" aria-labelledby="settings-section-user-account"
+        div class="pf-c-form__section-title" id="settings-section-user-account" aria-hidden="true" User account management zone
```
I think `h2` would look nicer here (well, for all sections), and semantically better?...

josemigallas replied:
I see what you mean but Patternfly disagrees: https://www.patternfly.org/v4/components/form/html#form-sections 🤷🏻

mayorova replied:
yeah, but https://www.patternfly.org/v4/components/form/html/#usage says that `.pf-c-form__section-title` is applicable to multiple elements, `h2` included :innocent: 

also: https://github.com/patternfly/patternfly/issues/3972

So, I think it's at our discretion...

josemigallas replied:
Fair enough, let's use `h2` then. And also `label` instead of `div`.

mayorova commented:
I don't really like the **Advanced Plans** section, as I think the hierarchy is really mixed up.

Like "Account plans changing" should be really "under" Account Plans section, and same for "Service Plan changing".

Also, 

"Change plan directly if developer has credit card, otherwise:" title is specifically for the two checkboxes:
- Only request a plan change
- Request Credit Card entry for paid plans

josemigallas commented:
I noticed the radiogroups are supposed to hid when the checkbox is unchecked, so I updated it. I also removed the "Advanced plans" section:

![Screenshot 2023-07-18 at 15 08 02](https://github.com/3scale/porta/assets/11672286/990d4c51-81ae-42bb-abd7-8fb47d58e663)

mayorova commented:
> I noticed the radiogroups are supposed to hid when the checkbox is unchecked, so I updated it. I also removed the "Advanced plans" section:

hmmmm, really? it's not like it works in production :shrug:

josemigallas commented on app/views/shared/_plan_change_settings.html.slim:
```diff
@@ -1,21 +1,21 @@
+  = render 'plan_change_radio', permission: permission,
+                                value: 'direct',
+                                checked: current_setting == 'direct',
+                                hint: can?(:see, :finance) ? t('formtastic.hints.settings.change_plan_finance') : nil
```
This is the lesser evil. Either we tweak PF4 in our favor or we start a css-wrestling battle agains it that eventually could break... OR we use a different design for this radiogroup.

mayorova commented:
@josemigallas 

I've tried these two options:

Account Plans are similar to what we have now (using `div class="pf-c-form__group-label-main"` for extra text)

Service Plans are with rewording, that hopefully makes the behavior more clear (and the structure more flat)

Basically, it's translating the logic that is defined [here](https://github.com/3scale/porta/blob/8bc4ef1d5ebf8e0794b2d5c684cccd3abd7b746c/app/lib/logic/plan_changes.rb#L77-L101) into human language.

josemigallas commented:
> Which one do you like more? (or maybe none? 😅 )

I like the second one a lot. Docs could take a look at the whole page's microcopy and these new radios to ensure the have the right wording.

mayorova commented on config/locales/en.yml:
```diff
@@ -481,6 +481,12 @@ en:
+        plans_info: |
+          The following options define whether the developers are allowed to change the plan
+          directly, or only request a plan change. When plan change is requested,
+          notifications about the requests will be sent to the admin users by email. Th plan
```
`The plan`

josemigallas replied:
```suggestion
          notifications about the requests will be sent to the admin users by email. The plan
```

jlledom commented on config/locales/en.yml:
```diff
@@ -1327,6 +1333,12 @@ en:
         public_search: 'Enable public search on Developer Portal'
         cms_escape_draft_html: 'Use Automatic Escaping for draft content'
         cms_escape_published_html: 'Use Automatic Escaping for published content'
+        change_plan_permission:
+          request: Request a plan change
+          direct: Change the plan directly
+          credit_card: Only request a plan change
+          request_credit_card: Request Credit Card entry for paid plans
+          none: Don't allow plan changes
```
After removing the hints (`"Change plan directly if developer has credit card, otherwise:"`), these options are not clear for me now. For instance, what's the difference between `"Request a plan change"` and `"Only request a plan change"`?

I think it needs some rewording. Just a suggestion:

```suggestion
          none: Don't allow plan changes
          request: Request a plan change
          direct: Change the plan directly
          credit_card: Change the plan if developer has CC, request a plan change otherwise
          request_credit_card: Change the plan if developer has CC, request CC details otherwise
```

mayorova replied:
Agreed, it's not clear at all :upside_down_face: 

I suggested this wording: https://github.com/3scale/porta/pull/3467#issuecomment-1641979386 (in Service Plans section). It is probably too long... but I think it's the only way to clarify the selected behavior.

jlledom replied:
Oh yes, true 👍.

josemigallas replied:
I missed this 😬. Here is the new wording: https://github.com/3scale/porta/pull/3509

mayorova commented:
Apart from the comments, it looks really nice!
